[Filebeat][New Module] Adding support for Oracle Audit logs

P1llus opened the pull request:
## What does this PR do?

This PR is the initial release of a Oracle module, with a audit log fileset.

## Why is it important?

Adding support for more OOTB integrations

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

elasticmachine commented:
Pinging @elastic/siem (Team:SIEM)

elasticmachine commented:
## :green_heart: Build Succeeded


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-21991/detail/PR-21991/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-21991/detail/PR-21991/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-21991/detail/PR-21991/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-21991/detail/PR-21991/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://beats_21991.docs-preview.app.elstc.co/diff)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats

* Build Cause: `[Pull request #21991 opened]`

* Start Time: 2020-10-19T16:21:57.000+0000

* Duration: 66 min 44 sec




#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 0  |
| Passed       | 4493  |
| Skipped      | 561 |
| Total        | 5054   |


</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY-->





<!--COMMENT_GENERATED_WITH_ID_comment.id-->

elasticmachine commented:
Pinging @elastic/integrations-services (Team:Services)

andresrc commented:
ping @sorantis @sayden

leehinman commented on x-pack/filebeat/module/oracle/database_audit/test/ORCLCDB_j002_28264_20201007122838056263426565.aud.log-expected.json:
```diff
@@ -0,0 +1,126 @@
+        "client.user.name": "",
+        "event.dataset": "oracle.database_audit",
+        "event.module": "oracle",
+        "event.timezone": "-02:00",
```
I think the original event had a timestamp of Wed Oct  7 12:28:38 2020 -04:00, so shouldn't the event.timezone be -04:00?

leehinman commented on x-pack/filebeat/module/oracle/database_audit/test/ORCLCDB_j002_28264_20201007122838056263426565.aud.log-expected.json:
```diff
@@ -0,0 +1,126 @@
+[
+    {
+        "@timestamp": "2020-10-07T16:28:38.000Z",
+        "client.address": "",
```
any chance we could not map empty strings?

leehinman commented on x-pack/filebeat/module/oracle/database_audit/ingest/pipeline.yml:
```diff
@@ -0,0 +1,118 @@
+    lang: painless
+    if: ctx?.oracle?.database_audit != null
+    source: |
+      void handleMap(Map map) {
```
Untested but something like this is a little closer to a classical recursive implementation:
```
void removeNulls(Object obj) {
    if (obj instance of Map) {
	for (def x : obj.entrySet()) {
	    if (x.getValue() == null) {
		obj.remove(x.getKey());
		continue;
	    }
	    removeNulls(x);
	}
    }
    if (obj instance of List) {
	for (def x : obj) {
	    removeNulls(x);
	}
    }
}
removeNulls(ctx);
```

P1llus replied:
After discussing a bit around handling both list and maps with @leehinman we decided to keep it as it is for now, might change in the future.

leehinman reviewed (changes requested):
Looking good.

Do you think we can add ECS categorization fields?

P1llus commented:
@leehinman 
For timezones I removed the default filebeat ones (removing add_locale processor), and I added a grok parser to grab the correct timezone since the date processor does not add timezone for some reason. If I don't set a timezone then event.timezone will stay empty, but it will be translated to UTC correctly.

For ECS Categorization I added event.type, event.cateogry, event.outcome and event.action. At this moment they really need to be static as there is no good way to differentiate the intent (is it a change, access, deletion etc).

The rest of the comments should also have been resolved with minor tweaks of the pipeline :)

elasticmachine commented:
## :green_heart: Flaky test report
Tests succeeded.


<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

### Test stats :test_tube:
| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 0  |
| Passed       | 4493  |
| Skipped      | 561 |
| Total        | 5054   |




</p>
</details>

<!--COMMENT_GENERATED_WITH_ID_flaky.id-->

leehinman reviewed (approved):
Changes LGTM

might want to set `event.kind = event`

sayden commented on x-pack/filebeat/module/oracle/database_audit/_meta/fields.yml:
```diff
@@ -0,0 +1,70 @@
+      description: >
+        Database Audit Status.
+
+    - name: sessionid
```
This should be `session.id` or even `session_id` if the field is (and will be) single field.

sayden commented on x-pack/filebeat/module/oracle/database_audit/_meta/fields.yml:
```diff
@@ -0,0 +1,70 @@
+      description: >
+        Session ID related to the audit event.
+
+    - name: client_terminal
```
`client.terminal` is better to encompass the next two under the `client` branch

sayden commented on x-pack/filebeat/module/oracle/database_audit/_meta/fields.yml:
```diff
@@ -0,0 +1,70 @@
+      description: >
+        If available, the client terminal type, for example "pty"
+
+    - name: client_address
```
`client.address`

sayden commented on x-pack/filebeat/module/oracle/database_audit/_meta/fields.yml:
```diff
@@ -0,0 +1,70 @@
+      description: >
+        The IP Address or Domain used by the client.
+
+    - name: client_user
```
`client.user`

sayden commented on x-pack/filebeat/module/oracle/database_audit/_meta/fields.yml:
```diff
@@ -0,0 +1,70 @@
+      description: >
+        The user running the client or connection to the database
+
+    - name: database_user
```
`database.user` probably better than `database_user` in case we improve the module in the future to extract more "database related" fields

sayden commented on x-pack/filebeat/module/oracle/database_audit/_meta/fields.yml:
```diff
@@ -0,0 +1,70 @@
+      description: >
+        The privilege group related to the database user
+
+    - name: entryid
```
`entry.id`

sayden commented on x-pack/filebeat/module/oracle/database_audit/_meta/fields.yml:
```diff
@@ -0,0 +1,70 @@
+      description: >
+        EntryID
+
+    - name: userhost
```
`user.host` but I was wondering if we can change this field to `database.host` according to the description or that will be awkward for Oracle users.

sayden commented on x-pack/filebeat/module/oracle/database_audit/_meta/fields.yml:
```diff
@@ -0,0 +1,70 @@
+      description: >
+        Action number related to the audit event
+
+    - name: dbid
```
`db.id` somehow this is asking me to be `database.id` to improve readability.

sayden commented on x-pack/filebeat/module/oracle/database_audit/_meta/fields.yml:
```diff
@@ -0,0 +1,70 @@
+    - name: length
+      type: long
+      description: >
+        The duration length.
```
Can we extend this description and maybe give the field a more meaningful name? I like to think that Beats modules are built for experts and non-experts alike so if you are not an expert, by reading the description you can understand why this field is important in your Kibana dashboard. And if you are not going to write this into a Kibana dashboard maybe the field isn't that important 😄 I'm not an expert so I have no clue why this field is important 😂

sayden reviewed (commented):
I'd change all `database_audit` mentions to `audit_logs` for example or simply `audit`. Maybe it's just me but the word `database` there feels redundant Maybe I'm missing that there are more type of audit logs and we are specifying which ones are those? Like the `database_audit` but you also have the `server_audit` and the `user_audit`? I don't know

P1llus commented:
> I'd change all `database_audit` mentions to `audit_logs` for example or simply `audit`. Maybe it's just me but the word `database` there feels redundant Maybe I'm missing that there are more type of audit logs and we are specifying which ones are those? Like the `database_audit` but you also have the `server_audit` and the `user_audit`? I don't know

Much appreciated feedback, thanks @sayden ! :+1: 

I think I agree with most of the field name changes, I was unsure if we wanted to keep the original names or not, because once we start renaming fields we have to ensure that any new fields that comes up at a later date does not conflict with the current ones.

On the topic of database_audit, we unfortunately have to keep some sort of reference to database here, as there is sooo many Oracle products, and a large amount of their portfolio has audit functionality, so we need to differentiate somewhat between products here I would say?

sayden commented:
The convention on Beats is to prettify everything with some exceptions in some specific fields (none that comes to my memory right now). It's not that I agree a lot, I think I'd prefer to maintain original names but this "prettifycation" is what you'll find in most Beats modules, so it's good for Beats users to maintain that familiarity between modules.

About the `database_audit` topic: sure! if you feel it's makes things more clear with the rest of oracle products that's great and thank you very much for the clarification and explanation :)

P1llus commented:
Should hopefully be good now @sayden . Always open for more changes if needed :+1:

sayden commented on x-pack/filebeat/module/oracle/database_audit/ingest/pipeline.yml:
```diff
@@ -0,0 +1,142 @@
+    ignore_missing: true
+- rename:
+    field: oracle.database_audit.client_user
+    target_field: client.user.name
```
Sorry but I didn't explicitly say it but I mean that the key should be `oracle.database_audit.client.user` I just didn't mention the beginning :sweat_smile: We cannot remove the module and fileset name unless we talk about ECS fields. So yeah, we need to maintain `oracle.database_audit.` to all extracted fields

sayden replied:
ECS fields :slightly_smiling_face:

sayden commented on x-pack/filebeat/module/oracle/database_audit/ingest/pipeline.yml:
```diff
@@ -0,0 +1,142 @@
+    ignore_missing: true
+- rename:
+    field: oracle.database_audit.userhost
+    target_field: server.address
```
`userhost` renamed to `server.address` Just double checking that this is ok

sayden replied:
ECS fields :slightly_smiling_face:

sayden commented on x-pack/filebeat/module/oracle/database_audit/ingest/pipeline.yml:
```diff
@@ -0,0 +1,142 @@
+    - "%{ISO8601_TIMEZONE:event.timezone}$"
+- rename:
+    field: oracle.database_audit.privilege
+    target_field: host.user.roles
```
Add `oracle.database_audit.` to the field name. as explained here https://github.com/elastic/beats/pull/21991/files/803b56530979a9192c2ffa306635e5328f9935bf#r517247123

sayden replied:
ECS fields :slightly_smiling_face:

sayden commented on x-pack/filebeat/module/oracle/database_audit/ingest/pipeline.yml:
```diff
@@ -0,0 +1,142 @@
+    ignore_missing: true
+- rename:
+    field: oracle.database_audit.client_address
+    target_field: client.address
```
Add `oracle.database_audit.` to the field name. as explained here https://github.com/elastic/beats/pull/21991/files/803b56530979a9192c2ffa306635e5328f9935bf#r517247123

sayden replied:
ECS fields :slightly_smiling_face:

sayden commented on x-pack/filebeat/module/oracle/database_audit/ingest/pipeline.yml:
```diff
@@ -0,0 +1,142 @@
+    ignore_missing: true
+- rename:
+    field: oracle.database_audit.database_user
+    target_field: server.user.name
```
Add `oracle.database_audit.` to the field name. as explained here https://github.com/elastic/beats/pull/21991/files/803b56530979a9192c2ffa306635e5328f9935bf#r517247123

sayden replied:
ECS fields :slightly_smiling_face:

sayden commented on x-pack/filebeat/module/oracle/database_audit/ingest/pipeline.yml:
```diff
@@ -0,0 +1,142 @@
+    field: oracle.database_audit.database_user
+    target_field: server.user.name
+    ignore_missing: true
+- convert:
```
You can use Grok to avoid the use of the `convert` processor here, like `%{PATTERN:fieldpath:type}` look at this example here https://github.com/elastic/beats/blob/master/filebeat/module/apache/access/ingest/pipeline.yml#L16

Sorry for not noticing this before :see_no_evil:

P1llus replied:
I think in terms of performance, the convert processor would be much quicker. Is there a reason why we would use GROK over convert when we only want to convert a single key-value field?

sayden replied:
No specific reasons apart from having a smaller pipeline. I haven't seen any benchmarks anyways but it's not a big deal, you can leave it. Just wanted to let you know in case you didn't know already :wink:

sayden commented:
Good work with the module!

elasticmachine commented:
Pinging @elastic/security-external-integrations (Team:Security-External Integrations)
